lib/resty/mlcache.lua

```diff
@@ -250,25 +292,58 @@ function _M.new(name, shm, opts)
         l1_serializer   = opts.l1_serializer,
     }
 
-    if opts.ipc_shm then
-        local mlcache_ipc = require "resty.mlcache.ipc"
+    if opts.ipc then
+        self.events = {
+            ["invalidation"] = {
+                channel = fmt("mlcache:invalidations:%s", name),
+                handler = function(key)
+                    self.lru:delete(key)
+                end,
+            },
+            ["purge"] = {
+                channel = fmt("mlcache:purge:%s", name),
+                handler = function()
+                    rebuild_lru(self)
+                end,
+            }
+        }
+
+        if opts.ipc.type == "mlcache_ipc" then
+            local mlcache_ipc = require "resty.mlcache.ipc"
+
+            local ipc, err = mlcache_ipc.new(opts.ipc.shm, opts.ipc.debug)
+            if not ipc then
+                return nil, "failed to initialize mlcache IPC " ..
+                            "(could not instantiate mlcache.ipc): " .. err
+            end
 
-        local err
-        self.ipc, err = mlcache_ipc.new(opts.ipc_shm, opts.debug)
-        if not self.ipc then
-            return nil, "could not instanciate mlcache.ipc: " .. err
-        end
+            for _, ev in pairs(self.events) do
+                ipc:subscribe(ev.channel, ev.handler)
+            end
+
+            self.broadcast = function(channel, data)
+                return ipc:broadcast(channel, data)
+            end
 
-        self.ipc_invalidation_channel = fmt("mlcache:invalidations:%s", name)
-        self.ipc_purge_channel = fmt("mlcache:purge:%s", name)
+            self.poll = function(timeout)
+                return ipc:poll(timeout)
+            end
+
+            self.ipc = ipc
 
-        self.ipc:subscribe(self.ipc_invalidation_channel, function(key)
-            self.lru:delete(key)
-        end)
+        elseif opts.ipc.type == "custom" then
+            local pok, perr = pcall(opts.ipc.register_listeners, self.events)
```

Comment:
This requires the callback function to throw errors and not return nil+err. That probably needs to be documented along with the callback signature.

Edit: since the other callbacks below are not called with a pcall, you might want to skip it here as well. Especially since this is initialization, so it wouldn't be an issue to break here if an uncaught error was thrown.

Reply:
Yeah, I initially thought it would be nice to catch Lua errors here, but on second thoughts, I have doubts. Other "initialization failures" throw errors - I think the best here is to accept `nil+err`, and to not run in `pcall` -> if an error is thrown, the stacktrace should lead to the user's function, because it is easier to track it. Will revert this and add soft error handling.
Thanks!

Reply:
(tbh I don't think I will extensively document the custom IPC thing, it is mostly for "advanced" users imho)

Reply:
I somewhat changed my mind, and went with keeping the `opts.ipc_shm` option to enable the off the shelf IPC module of this library, and allowing custom IPC via `opts.ipc` - I think it is the best of both worlds. Thanks for the suggestions!